src/main/java/com/example/service/impl/ShoppingCartServiceImpl.java

```diff
@@ -0,0 +1,98 @@
+package com.example.service.impl;
+
+import com.example.dto.cartitem.CartItemResponseDto;
+import com.example.dto.cartitem.CreateCartItemRequestDto;
+import com.example.dto.shoppingcart.ShoppingCartResponseDto;
+import com.example.dto.shoppingcart.UpdateCartItemRequestDto;
+import com.example.exception.BookAlreadyAddedToCartException;
+import com.example.exception.CartItemAbsenceException;
+import com.example.exception.EntityNotFoundException;
+import com.example.mapper.CartItemMapper;
+import com.example.mapper.ShoppingCartMapper;
+import com.example.model.CartItem;
+import com.example.model.ShoppingCart;
+import com.example.repository.book.BookRepository;
+import com.example.repository.cartitem.CartItemRepository;
+import com.example.repository.shoppingcartrepository.ShoppingCartRepository;
+import com.example.service.ShoppingCartService;
+import lombok.RequiredArgsConstructor;
+import org.springframework.stereotype.Service;
+
+@Service
+@RequiredArgsConstructor
+public class ShoppingCartServiceImpl implements ShoppingCartService {
+    private final ShoppingCartRepository shoppingCartRepository;
+    private final ShoppingCartMapper shoppingCartMapper;
+    private final CartItemRepository cartItemRepository;
+    private final CartItemMapper cartItemMapper;
+    private final BookRepository bookRepository;
+
+    @Override
+    public ShoppingCartResponseDto getByUserId(Long userId) {
+        return shoppingCartMapper.toDto(getShoppingCart(userId));
+    }
+
+    @Override
+    public ShoppingCartResponseDto addCartItem(Long userId, CreateCartItemRequestDto requestDto) {
+        ShoppingCart shoppingCart = getShoppingCart(userId);
+        CartItem cartItem = cartItemMapper.toModel(requestDto);
+        checkIfBookExist(requestDto);
+
+        if (shoppingCart.getCartItems().stream()
+                .anyMatch(item -> item.getBook().getId().equals(requestDto.getBookId()))) {
+            throw new BookAlreadyAddedToCartException("Book already added in shopping cart");
```

Comment:
In real life when we add some product in shopping cart, the quantity just increased. 
So we can implement this logic
```
check if book exists in DB
ShoppingCart cart = repository.findByUserId(userId);
cart.getCartItems().stream()
                .filter(item -> item.getBook().getId().equals(requestDto.bookId()))
                .findFirst()
                .ifPresentOrElse(item -> item.setQuantity(item.getQuantity() + requestDto.quantity()),
                        () -> addCartItemToCart(requestDto, book, cart));
cartItemRepository.save(cartItem);
return shoppingCartMapper.toDto(cart);
```

Where 
```
private void addCartItemToCart(CartItemRequestDto requestDto, Book book, ShoppingCart cart) {
        CartItem cartItem = cartItemMapper.toCartItem(requestDto);
        cartItem.setBook(book);
        cart.addItemToCart(cartItem);
    }
```

Where in ShoppingCart class
```
public void addItemToCart(CartItem cartItem) {
        cartItem.setShoppingCart(this);
        cartItems.add(cartItem);
    }
```

Reply:
done